src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -16,32 +18,47 @@ internal class EnumConverter<T> : JsonConverter<T>
         // Odd type codes are conveniently signed types (for enum backing types).
         private static readonly string? s_negativeSign = ((int)s_enumTypeCode % 2) == 0 ? null : NumberFormatInfo.CurrentInfo.NegativeSign;
 
+        private const string ValueSeparator = ", ";
+
         private readonly EnumConverterOptions _converterOptions;
-        private readonly JsonNamingPolicy _namingPolicy;
-        private readonly ConcurrentDictionary<string, string>? _nameCache;
+        private readonly JsonNamingPolicy? _namingPolicy;
+        private readonly ConcurrentDictionary<ulong, JsonEncodedText> _nameCache;
 
         public override bool CanConvert(Type type)
         {
             return type.IsEnum;
         }
 
-        public EnumConverter(EnumConverterOptions options)
-            : this(options, namingPolicy: null)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonSerializerOptions serializerOptions)
+            : this(converterOptions, namingPolicy: null, serializerOptions)
         {
         }
 
-        public EnumConverter(EnumConverterOptions options, JsonNamingPolicy? namingPolicy)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? namingPolicy, JsonSerializerOptions serializerOptions)
         {
-            _converterOptions = options;
-            if (namingPolicy != null)
-            {
-                _nameCache = new ConcurrentDictionary<string, string>();
-            }
-            else
+            _converterOptions = converterOptions;
+            _namingPolicy = namingPolicy;
+            _nameCache = new ConcurrentDictionary<ulong, JsonEncodedText>();
+
+            string[] names = Enum.GetNames(TypeToConvert);
+            Array values = Enum.GetValues(TypeToConvert);
+            Debug.Assert(names.Length > 0 && names.Length == values.Length);
+
+            JavaScriptEncoder? encoder = serializerOptions.Encoder;
+
+            for (int i = 0; i < names.Length; i++)
             {
-                namingPolicy = JsonNamingPolicy.Default;
+                T value = (T)values.GetValue(i)!;
+                // Enum values can be represented as ulong. Note that F# supports char as enum backing types.
+                ulong key = Unsafe.As<T, ulong>(ref value);
```

Comment:
Don't use `Unsafe.As` here: you could buffer overrun. Use `Convert.ToUInt64(value, CultureInfo.InvariantCulture)` instead.

Reply:
As discussed offline, I adapted the following code to properly do this https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Private.CoreLib/src/System/Enum.cs#L240-L260